### PR TITLE
perf: use diesel_async and improve queries

### PR DESCRIPTION
**Pull request by lajp**
Diesel has some issue on complex queries and therefore we have to rely on multiple queries in some cases.

Regardless of that this is still a massive improvement on the current implementation.
For example the endpoint `/users/@me/leaderboards` is sped up roughly by a factor of 37.
## before:
image
## after:
image

The performance of `/friends/list` is also drastically improved.

### CODE REVIEW

**Comment by lajp** [comment]
This is now ready for review.

File: src/database/friends.rs [hunk 1]
```diff
@@ -91,49 +64,61 @@ impl super::DatabaseWrapper {
...
+                            } else {
+                                0
+                            }
+                        })
+                        .sum(),
+                    past_week: d
+                        .iter()
+                        .map(|a| {
+                            if a.start_time
+                                >= chrono::Local::now().naive_local() - chrono::Duration::days(30)
```
Comment (Eldemarkki) [inline src/database/friends.rs:109]:
```suggestion
                                >= chrono::Local::now().naive_local() - chrono::Duration::days(7)
```

**Review by Eldemarkki — CHANGES_REQUESTED** [review]

File: src/database/friends.rs (hunk 1 above)
Comment (lajp) [reply src/database/friends.rs:109]: Nice catch!

File: src/database/auth.rs [hunk 2]
```diff
@@ -3,94 +3,104 @@ use argon2::{
...
             Ok(None)
         }
     }
 
     pub async fn regenerate_token(&self, userid: i32) -> Result<String, TimeError> {
+        let mut conn = self.db.get().await?;
+
         let token = crate::utils::generate_token();
 
         let token_clone = token.clone();
```
Comment (Eldemarkki) [inline src/database/auth.rs:96]: I'm no expert in Rust so feel free to disregard if this is invalid, but is this clone necessary anymore now that we don't have the `move |mut conn| { ... }`?

File: src/database/auth.rs [hunk 3]
```diff
@@ -224,24 +253,20 @@ impl super::DatabaseWrapper {
 
         let user_id_arg_clone = user_id_arg.clone();
```
Comment (Eldemarkki) [inline src/database/auth.rs:254]: Same thing here, is this clone necessary?

File: src/database/auth.rs (hunk 2 above)
Comment (lajp) [reply src/database/auth.rs:96]: You're right it's not required.

File: src/database/auth.rs (hunk 3 above)
Comment (lajp) [reply src/database/auth.rs:254]: It's not, thanks

**Review by Eldemarkki — APPROVED** [review]

**Review by DrVilepis — APPROVED** [review]
lgtm (?)

**Pull request merged**